cli/connector/eval_plain_text.go

```diff
@@ -0,0 +1,413 @@
+		if err == io.EOF {
+			return nil, err
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("Failed to read from instance socket: %s", err)
+		}
+
+		dataPortion := string(dataPortionBytes)
+
+		if !pushTagIsReceived(dataPortion) {
+			dataBytes = dataPortionBytes
+			break
+		}
+
+		if opts.PushCallback != nil {
+			var pushedData interface{}
+
+			pushedData, err := getPushedData(dataPortionBytes)
+			if err != nil {
+				return nil, err
+			}
+
+			opts.PushCallback(pushedData)
+		}
+	}
+
+	return dataBytes, nil
+}
+
+func readDataPortionFromPlainTextConn(conn net.Conn, readTimeout time.Duration) ([]byte, error) {
+	tmp := make([]byte, 1)
+	data := make([]byte, 0)
+
+	if readTimeout > 0 {
+		conn.SetReadDeadline(time.Now().Add(readTimeout))
+	} else {
+		conn.SetReadDeadline(time.Time{})
+	}
+
+	hasYAMLOutputPrefix := false
+
+	for {
+		if n, err := conn.Read(tmp); err != nil && err != io.EOF {
+			return nil, fmt.Errorf("Failed to read: %s", err)
+		} else if n == 0 || err == io.EOF {
+			return nil, io.EOF
+		} else {
+			data = append(data, tmp[:n]...)
+
+			dataString := string(data)
+			if strings.HasPrefix(endOfYAMLOutput, dataString) ||
+				strings.HasPrefix(tagPushPrefixYAML, dataString) ||
+				strings.HasPrefix(tagPushPrefixLua, dataString) {
+				continue
+			}
+
+			if !hasYAMLOutputPrefix &&
+				strings.HasPrefix(dataString, startOfYamlOutput) || strings.HasPrefix(dataString, tagPushPrefixYAML) {
+				hasYAMLOutputPrefix = true
+			}
+
+			if hasYAMLOutputPrefix && strings.HasSuffix(dataString, endOfYAMLOutput) {
+				break
+			}
+
+			if strings.HasSuffix(dataString, endOfLuaOutput) {
+				break
+			}
+		}
+	}
+
+	if len(data) == 0 {
+		return nil, fmt.Errorf("Connection was closed")
+	}
+
+	return data, nil
+}
+
+func pushTagIsReceived(dataPortion string) bool {
+	if strings.HasPrefix(dataPortion, tagPushPrefixYAML) {
+		return true
+	}
+
+	if strings.HasPrefix(dataPortion, tagPushPrefixLua) {
+		return true
+	}
+
+	return false
+}
+
+func getPushedData(pushedDataBytes []byte) (interface{}, error) {
+	var pushedData interface{}
+	pushedDataString := string(pushedDataBytes)
+
+	if strings.HasPrefix(pushedDataString, tagPushPrefixYAML) {
+		// YAML - just decode tag
+		if err := yaml.Unmarshal(pushedDataBytes, &pushedData); err != nil {
+			return nil, fmt.Errorf("Failed to unmarshal pushed data: %s", err)
+		}
+	} else {
+		// Lua
+
+		// remove first line (-- Push)
+		pushedDataString = strings.SplitN(pushedDataString, "\n", 2)[1]
+		// remove ";"
+		pushedDataString = strings.TrimRight(pushedDataString, ";")
+
+		pushedData = pushedDataString
+	}
+
+	return pushedData, nil
+}
+
+func processEvalTarantoolRes(resBytes []byte, result interface{}) ([]interface{}, error) {
+	var err error
+	var evalResultEncHex string
+
+	var getResultEncHexFunc func([]byte) (string, error)
+	// Result data is returned as a table
+	// `{ data_enc = msgpack.encode(ret):hex() }`.
```

Comment:
Could you eliminate "hex" here?

Anyway it's an array of bytes.

Reply:
There is a problem with Lua output. The approach with hex encoding is the simplest now, I believe it become better sometimes.

```
tarantool> res = require('console').eval('return true')
;
tarantool> res
"true;";
tarantool> msgpack.encode(res)
"�true;";
```